core/src/main/java/hudson/model/queue/WorkUnitContext.java

```diff
@@ -108,12 +108,17 @@ public WorkUnit getPrimaryWorkUnit() {
      * All the {@link Executor}s that jointly execute a {@link Task} call this method to synchronize on the start.
      */
     public void synchronizeStart() throws InterruptedException {
-        startLatch.synchronize();
-        // the main thread will send a notification
-        Executor e = Executor.currentExecutor();
-        WorkUnit wu = e.getCurrentWorkUnit();
-        if (wu.isMainWork()) {
-            future.start.set(e.getCurrentExecutable());
+        // Let code waiting for the start (future.start.get()) finishes when there is a faulty SubTask by setting the
+        // future always.
+        try {
+            startLatch.synchronize();
+        } finally {
+            // the main thread will send a notification
+            Executor e = Executor.currentExecutor();
+            WorkUnit wu = e.getCurrentWorkUnit();
+            if (wu.isMainWork()) {
+                future.start.set(e.getCurrentExecutable());
+            }
         }
     }
 
```

Comment:
We should block out any direct use of this method, `DoNotUse.class` might be appropriate

```suggestion

    @Restricted(NoExternalUse.class)
```